core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -0,0 +1,204 @@
+                        }
+
+                        startTime = System.nanoTime();
+                    }
+                }
+            } catch (IOException e) {
+                logger.log(Level.WARNING, e.getMessage(), e);
+            } catch (InterruptedException e) {
+                logger.info("Shutting down camera server");
+                serverThread = Optional.empty();
+                return;
+            } finally {
+                try {
+                    socket.close();
+                } catch (Throwable ignored) {
+                }
+
+                connected = false;
+            }
+        }
+    };
+
+    @Override
+    public String getName() {
+        return "Publish Video";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Publish an M-JPEG stream to the dashboard";
+    }
+
+    @Override
+    public Optional<InputStream> getIcon() {
+        return Optional.of(getClass().getResourceAsStream("/edu/wpi/grip/ui/icons/publish-video.png"));
+    }
+
+    @Override
+    public InputSocket<?>[] createInputSockets(EventBus eventBus) {
+        eventBus.register(this);
+        return new InputSocket<?>[]{
+                new InputSocket<>(eventBus, SocketHints.Inputs.createMatSocketHint("Image", false)),
+                new InputSocket<>(eventBus, SocketHints.Inputs.createNumberSliderSocketHint("Quality", 80, 0, 100)),
+        };
+    }
+
+    @Override
+    public OutputSocket<?>[] createOutputSockets(EventBus eventBus) {
+        return new OutputSocket<?>[0];
+    }
+
+    @Override
+    public Optional<?> createData() {
+        numSteps++;
+        if (!serverThread.isPresent()) {
+            serverThread = Optional.of(new Thread(runServer));
+            serverThread.get().start();
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public void perform(InputSocket<?>[] inputs, OutputSocket<?>[] outputs) {
+        Mat input = (Mat) inputs[0].getValue().get();
+        Number quality = (Number) inputs[1].getValue().get();
+
+        if (!connected) {
+            return; // Don't waste any time converting images if there's no dashboard connected
+        }
+
+        if (input.empty()) {
+            throw new IllegalArgumentException("Input image must not be empty");
+        }
+
+        FrameConverter<Mat> frameConverter = new OpenCVFrameConverter.ToMat();
+        FrameConverter<BufferedImage> imageConverter = new Java2DFrameConverter();
+        BufferedImage image = imageConverter.convert(frameConverter.convert(input));
+
+        synchronized (imageLock) {
+            this.image = image;
+            this.compressionQuality = quality.floatValue();
+            imageLock.notify();
+        }
+    }
+
+    @Subscribe
+    public void onStepRemoved(StepRemovedEvent event) {
+        if (event.getStep().getOperation() == this) {
+            if (--numSteps == 0) {
+                serverThread.ifPresent(Thread::interrupt);
```

Comment:
Make sure that this thread joins correctly. You can do this in a separate thread if you want to keep this from hanging.